qiskit/qpy/binary_io/value.py

```diff
@@ -217,32 +216,45 @@ def dumps_value(obj):
     Raises:
         QpyError: Serializer for given format is not ready.
     """
-    type_key = TypeKey.assign(obj)
+    type_key = type_keys.Value.assign(obj)
 
-    if type_key == TypeKey.INTEGER:
+    if type_key == type_keys.Value.INTEGER:
         binary_data = struct.pack("!q", obj)
-    elif type_key == TypeKey.FLOAT:
+    elif type_key == type_keys.Value.FLOAT:
         binary_data = struct.pack("!d", obj)
-    elif type_key == TypeKey.COMPLEX:
+    elif type_key == type_keys.Value.COMPLEX:
         binary_data = struct.pack(formats.COMPLEX_PACK, obj.real, obj.imag)
-    elif type_key == TypeKey.NUMPY_OBJ:
+    elif type_key == type_keys.Value.NUMPY_ARR:
+        binary_data = obj.tobytes()
```

Comment:
This isn't safe to do, the raw memory is missing the necessary metadata that npy (ie `np.save()`) uses to recreate the array. You'll have issues related to the in RAM layout of the array between systems using `tobytes()`.

Reply:
Yes, I noticed this and removed. However, I ran into failure in waveform serialization.
```python
import io
import numpy as np

t = np.linspace(0, 1, 100)
waveform = 0.1 * np.sin(2 * np.pi * t)

file = io.BytesIO()
np.save(file, waveform)
np.load(file)
```
this results in
```
ValueError: Cannot load file containing pickled data when allow_pickle=False
```
Setting `allow_pickle=True` doesn't work. I also found `tobytes` doesn't save metadata, but it works for waveform samples because data is always 1d complex ndarray. So I removed `NUMPY_ARR` format, but waveform sample serialization still uses `tobytes`.

Reply:
Thanks Matthew it worked! https://github.com/Qiskit/qiskit-terra/pull/7300/commits/5c45c11fb76f4609d4817f0c3ac35bde0ae0d5c5